---
title: '`r rstudioapi::getSourceEditorContext()$path %>% basename() %>% gsub(pattern = "\\.Rmd",replacement = "")`' 
author: "Avishai Wizel"
date: '`r Sys.time()`'
output:
  # html_notebook: 
  #   code_folding: hide
  #   toc: true
  #   toc_collapse: true
  #   toc_float:
  #     collapsed: false
  #   number_sections: true
  #   toc_depth: 1
  html_document: 
    code_folding: hide
    toc: yes
    toc_collapse: yes
    toc_float: 
      collapsed: TRUE
    number_sections: true
    toc_depth: 2
    df_print: paged
params: 
  data_out_dir: "./temp_data/" 
  input_data: !r c(
    hmsc_cancer_cells = "./Reports/HMSC/01_preprocess/03_HMSC_cells_preprocess/acc1_cancer_cells_2500features_integrated_V5.RDS",
    genesets_h = "./Input_data/h.all.v7.0.symbols.pluscc.gmt",
    global_params  = "./Input_data/global_params.R")
---



# Functions

```{r warning=FALSE}
source_from_github(repositoy = "DEG_functions",version = "0.2.54")
source_from_github(repositoy = "HMSC_functions",version = "0.1.14",script_name = "functions.R")
source_from_github(repositoy = "sc_general_functions",version = "0.1.34",script_name = "functions.R")
library(hypeR)
source(params$input_data["global_params"])
```

# Data

```{r}
hmsc_cancer_cells = readRDS(params$input_data["hmsc_cancer_cells"])
genesets_h  = getGmt(params$input_data["genesets_h"])
pid_genesets = getGmt("./Input_data/c2.cp.pid.v2024.1.Hs.symbols.gmt")
```

# HPV UMAP

```{r}
HPV33_P3 = fread("./Data/HPV33_P3.txt",col.names = c("plate","reads")) %>% as.data.frame()
HPV33_P3.df = HPV33_P3 %>% mutate(
  plate = gsub(x =HPV33_P3$plate, replacement = "",pattern = "_.*$") 
  %>% gsub(pattern = "-P",replacement = ".P") 
  %>% gsub(pattern = "-",replacement = "_",)
  )
HPV33_P3.df = HPV33_P3.df %>% dplyr::filter(HPV33_P3.df$plate %in% colnames(hmsc_cancer_cells))
rownames(HPV33_P3.df)  <- HPV33_P3.df$plate
HPV33_P3.df$plate = NULL


HPV33_P2 = fread("./Data/HPV33_P2.txt",col.names = c("plate","reads")) %>% as.data.frame()
HPV33_P2.df = HPV33_P2 %>% mutate(
  plate = gsub(x =HPV33_P2$plate, replacement = "",pattern = "_.*$") 
  %>% gsub(pattern = "plate2-",replacement = "plate2_",)
  %>% gsub(pattern = "-",replacement = "\\.",)
  )
HPV33_P2.df = HPV33_P2.df %>% dplyr::filter(HPV33_P2.df$plate %in% colnames(hmsc_cancer_cells))
rownames(HPV33_P2.df)  <- HPV33_P2.df$plate
HPV33_P2.df$plate = NULL

HPV33 = rbind(HPV33_P3.df,HPV33_P2.df)
hmsc_cancer_cells = AddMetaData(object = hmsc_cancer_cells,metadata = HPV33,col.name = "HPV33.reads")
FeaturePlot(hmsc_cancer_cells,features = "HPV33.reads",max.cutoff = 10)
```

# HPV reads count
```{r}

data = FetchData(object = hmsc_cancer_cells,vars = "HPV33.reads")

data = data %>% mutate("0 reads" = if_else(condition = HPV33.reads == 0,true = 1,false = 0))
data = data %>% mutate("1 reads" = if_else(condition = HPV33.reads == 1,true = 1,false = 0))
data = data %>% mutate("2 reads" = if_else(condition = HPV33.reads == 2,true = 1,false = 0))
data = data %>% mutate("3-23 reads" = if_else(condition = HPV33.reads >=3 &HPV33.reads  <24,true = 1,false = 0))
data = data %>% mutate("24+ reads" = if_else(condition = HPV33.reads >=24,true = 1,false = 0))
data = colSums(data[,2:ncol(data)]) %>% as.data.frame()
names(data)[1] = "count"
data = rownames_to_column(data,var = "bin")
data
ggplot(data=data, aes(x=factor(bin,levels = c("0 reads","1 reads","2 reads","3-23 reads","24+ reads")), y=count)) +
  geom_bar(stat="identity", fill="steelblue") + xlab("HPV Reads")+ theme_minimal()+
  geom_text(aes(label=count), vjust=-0.5, color="black", size=3.5)
```

# MYB comparision
```{r}
hpv33_positive = HPV33 %>% dplyr::mutate(hpv33_positive = case_when(reads >= 10 ~ "positive",
                                                                    reads < 10 ~ "negative")
)



hpv33_positive$reads = NULL
hmsc_cancer_cells = AddMetaData(object = hmsc_cancer_cells,metadata = hpv33_positive)
```

```{r fig.height=6, fig.width=12}
DimPlot(object = hmsc_cancer_cells,group.by  = c("hpv33_positive"),pt.size = 2)+
FeaturePlot(object = hmsc_cancer_cells,features = "MYB",pt.size = 2)

```



# Fisher test
```{r}
DefaultAssay(hmsc_cancer_cells) = "RNA" # calculate MYB positive/negtive from original data
# set MYB pos/neg
hmsc_cancer_cells$MYB_status = FetchData(object = hmsc_cancer_cells,vars = "MYB") %>%
  mutate(MYB_status = if_else(
    condition = MYB > 0,
    true = "MYB_positive",
    false = "MYB_negative"
  )) %>% 
  pull(MYB_status)

DefaultAssay(hmsc_cancer_cells) = "integrated"

```

```{r }
# plot
library(rstatix)

df  = FetchData(object = hmsc_cancer_cells,
                vars = c("MYB_status", "hpv33_positive"))  %>%
  droplevels()
test = fisher_test(table(df))
p = ggbarstats(
  df,
  MYB_status,
  hpv33_positive,
  results.subtitle = FALSE,
  subtitle = paste0("Fisher's exact test", ", p-value = ",
                    test$p)
)

p
```
# MYB targets

```{r}
# load MYB targets
library("readxl")
myb_tergets <- read_excel("./Input_data/MYB_tergets_PMC4767593_sup_table_4.xlsx",col_names = T) %>% 
  pull(Gene)
```

```{r}

# add score
myb_tergets_score = FetchData(object = hmsc_cancer_cells,vars = myb_tergets,slot = "data") %>% rowMeans()
myb_tergets_df = data.frame(myb_tergets_score,row.names = names(myb_tergets_score))
hmsc_cancer_cells = AddMetaData(object = hmsc_cancer_cells,metadata = myb_tergets_df)

```


```{r}

genes_by_tp = FetchData(object = hmsc_cancer_cells,vars = c("hpv33_positive", "myb_tergets_score"),slot = "data") 

# all cells by patient boxplot
formula <- as.formula("myb_tergets_score ~ hpv33_positive")

#plot and split by patient:
stat.test <- genes_by_tp %>%
  t_test(formula) %>%
  adjust_pvalue(method = "fdr") %>%
  add_significance() %>%
  add_xy_position(x = "hpv33_positive") %>%
  mutate(p.adj = signif(p.adj, digits = 2))
stat.test


plt = ggplot(genes_by_tp, aes(x = hpv33_positive, y = myb_tergets_score)) +
  geom_violin(scale = "width",aes(fill = hpv33_positive)) +
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  stat_pvalue_manual(stat.test, label = "{p.adj}")  + #add p value
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) + # Add 10% spaces between the p-value labels and the plot border
  geom_jitter()+
  ggtitle("MYB targets")
plt
```

# DEG LR latent vars plate


```{r}
DefaultAssay(hmsc_cancer_cells) = "integrated"
hmsc_cancer_cells = SetIdent(hmsc_cancer_cells,value = "hpv33_positive")
```



```{r echo = T, results = 'hide'}

hmsc_cancer_cells = FindVariableFeatures(hmsc_cancer_cells,nfeatures = 20000,assay = "RNA")
var_features = VariableFeatures(hmsc_cancer_cells,assay = "RNA")
top_expressed_features = hmsc_cancer_cells@assays$RNA@data %>% rowMeans() %>% sort(decreasing = T) %>% head(3000) %>% names()
features = intersect(var_features, top_expressed_features)


acc_deg <-
  FindMarkers(
    hmsc_cancer_cells,
    ident.1 = "positive",
    ident.2 = "negative",
    features = features,
    densify = T,
    assay = "RNA",
    test.use = "LR",
    latent.vars = "plate",
    logfc.threshold = 0.1,
    min.pct = 0.1,
    mean.fxn = function(x) {
      return(log2(rowMeans(x)+1)) # change func to calculate logFC in log space data (default to exponent data)
    }
  )
acc_deg$fdr<-p.adjust(p = as.vector(acc_deg$p_val) ,method = "fdr")



```



```{r}
acc_deg$score = acc_deg$avg_log2FC*-log10(acc_deg$fdr)
ranked_vec = acc_deg[,"avg_log2FC"]%>% setNames(rownames(acc_deg)) %>% na.omit() # make named vector

genesets =geneIds(genesets_h)
hyp_obj <-hypeR_fgsea(signature = ranked_vec,genesets = genesets,up_only = F)
p = hyp_dots(hyp_obj,merge = F)
p[[1]]
p[[2]]
```
```{r}
pdf(paste0(params$data_out_dir,"HMSC_HPVon_HPVoff_GSEA.pdf"))
p[[1]]
dev.off()


```


```{r paged.print=TRUE}

paged_table(acc_deg[c("MYB","JAG1","MYC"),])
paged_table(acc_deg[myb_tergets, ])
```





volcano plot log2(mean logTPM HPV+/mean logTPM HPV-)
```{r}
plt = volcano_plot(
  de_genes = acc_deg,
  fc_cutoff = volcano_fc,
  fdr_cutoff =volcano_fdr,
  ident1 = "HPV33 positive",
  ident2 = "HPV33 negative",
  top_genes_text = 10,show_gene_names = c("MYB","JAG1","MYC")
) + ggtitle("Differential Gene Expression by HPV Status
")
plt
```




```{r}
pdf(paste0(params$data_out_dir,"HMSC_HPVon_HPVoff_volcano.pdf"))
plt
dev.off()


```





## Cycling cells filtering {.tabset}

```{r}
hallmark_name = "GO_MITOTIC_CELL_CYCLE"
hmsc_cancer_cells = FindVariableFeatures(object = hmsc_cancer_cells,nfeatures = 15000,assay = "RNA")
hmsc_cancer_cells = ScaleData(object = hmsc_cancer_cells,features = VariableFeatures(hmsc_cancer_cells,assay = "RNA"))
geneIds= genesets_h[[hallmark_name]]@geneIds %>% intersect(VariableFeatures(hmsc_cancer_cells,assay = "RNA")) 
score <- apply(hmsc_cancer_cells@assays$integrated@scale.data[geneIds,],2,mean)
hmsc_cancer_cells=AddMetaData(hmsc_cancer_cells,score,hallmark_name)
```

```{r echo=TRUE, results='asis'}
hmsc_cc_scores = FetchData(object = hmsc_cancer_cells,vars = "GO_MITOTIC_CELL_CYCLE")

distributions_plt = ggplot() +
  geom_density(aes(GO_MITOTIC_CELL_CYCLE, fill = "HMSC"), alpha = .2, data = hmsc_cc_scores) +
  scale_fill_manual(name = "Dataset", values = c(ACC = "red", HMSC = "green"))+ geom_vline(aes(xintercept=0.3),
            color="blue", linetype="dashed", size=1) +ggtitle("'GO_MITOTIC_CELL_CYCLE'  score distribution")

print_tab(plt = distributions_plt,title = "score distribution",subtitle_num = 3)

```

```{r warning=FALSE}


#filter:
hmsc_cancer_cells_ccFiltered=hmsc_cancer_cells[,hmsc_cancer_cells@meta.data[[hallmark_name]]< 0.3]


```


```{r}
DefaultAssay(hmsc_cancer_cells_ccFiltered) = "integrated"
hmsc_cancer_cells_ccFiltered = SetIdent(hmsc_cancer_cells_ccFiltered,value = "hpv33_positive")
```



```{r echo = T, results = 'hide'}
hmsc_cancer_cells = FindVariableFeatures(hmsc_cancer_cells,nfeatures = 20000,assay = "RNA")
var_features = VariableFeatures(hmsc_cancer_cells,assay = "RNA")


acc_deg <-
  FindMarkers(
    hmsc_cancer_cells_ccFiltered,
    ident.1 = "positive",
    ident.2 = "negative",
    features = var_features,
    densify = T,
    assay = "RNA",
    test.use = "LR",
    latent.vars = "plate",
    logfc.threshold = 0.1,
    min.pct = 0.1,
    mean.fxn = function(x) {
      return(log2(rowMeans(x)+1)) # change func to calculate logFC in log space data (default to exponent data)
    }
  )
acc_deg$fdr<-p.adjust(p = as.vector(acc_deg$p_val) ,method = "fdr")



```

```{r}
write.csv(x  = acc_deg,file = paste0(params$data_out_dir,"HMSC_hpv_deg_df_noG2M_cells.csv"),row.names = T)
```

```{r}
acc_deg$score = acc_deg$avg_log2FC*-log10(acc_deg$fdr)
ranked_vec = acc_deg[,"avg_log2FC"]%>% setNames(rownames(acc_deg)) %>% na.omit() # make named vector

genesets =geneIds(genesets_h)
hyp_obj <-hypeR_fgsea(signature = ranked_vec,genesets = genesets,up_only = F)
p = hyp_dots(hyp_obj,merge = F)
p[[1]]
p[[2]]
```

```{r}
pdf(paste0(params$data_out_dir,"HMSC_HPVon_HPVoff_GSEA_noG2M_cells.pdf"))
p[[1]]
dev.off()


```

# HPV vs genes
```{r fig.width=8}
top_genes = acc_deg %>% filter(.$avg_log2FC>0) %>% arrange(.$fdr) %>% head(5) %>% rownames()
top_genes = c(top_genes,c("MYB","JAG1"))

myb_vs_hpv = FetchData(object = hmsc_cancer_cells,vars = c("hpv33_positive",top_genes))
df = reshape2::melt(myb_vs_hpv,value.name = "Expression") %>% dplyr::rename(gene = variable)

library(rstatix)
stat.test <- df %>%
    group_by(gene) %>%
  wilcox_test(Expression ~ hpv33_positive) %>%
  mutate(y.position = 5)

stat.test

stat.test <- stat.test %>% 
  add_xy_position(x = "gene", dodge = 0.8)

ggboxplot(
  df,
  x = "gene",
  y = "Expression",
  color = "hpv33_positive",
  palette = "jco",
  add = "jitter"
)+ stat_pvalue_manual(stat.test,y.position = 13, label = "p = {p}",remove.bracket = T)
```
# OPSCC signature

```{r}
avg_log2FC_cutoff = log2(1.1)
fdr_cutoff = 0.05


opscc_deg = readRDS(file = "./Reports/HPV_OPSCC_Analysis_PMC10191634/02_findDEG/opscc_deg_5Kvargenes.rds")
hmsc_deg = read.table(file = "./Reports/HMSC/05_HPV_analysis/HMSC_hpv_deg_df.csv",row.names = 1,sep = ",",header = T)

hmsc_signatue = rownames(hmsc_deg[hmsc_deg$fdr<fdr_cutoff & hmsc_deg$avg_log2FC>avg_log2FC_cutoff,])


opscc_signature = opscc_deg %>% filter(avg_log2FC > avg_log2FC_cutoff &
                                  fdr < fdr_cutoff) %>% rownames()
opscc_signature
```

```{r}

# add score
opscc_hpv_score = FetchData(object = hmsc_cancer_cells,vars = opscc_signature,slot = "data") %>% rowMeans()
hpv_signature = data.frame(opscc_hpv_score,row.names = names(opscc_hpv_score))
hmsc_cancer_cells = AddMetaData(object = hmsc_cancer_cells,metadata = hpv_signature)

```




```{r}

genes_by_tp = FetchData(object = hmsc_cancer_cells,vars = c("hpv33_positive", "opscc_hpv_score"),slot = "data") 

# all cells by patient boxplot
formula <- as.formula("opscc_hpv_score  ~ hpv33_positive")

#plot and split by patient:
stat.test <- genes_by_tp %>%
  t_test(formula) %>%
  adjust_pvalue(method = "fdr") %>%
  add_significance() %>%
  add_xy_position(x = "hpv33_positive") %>%
  mutate(p.adj = signif(p.adj, digits = 2))
stat.test


plt = ggplot(genes_by_tp, aes(x = hpv33_positive, y = opscc_hpv_score )) +
  geom_violin(scale = "width",aes(fill = hpv33_positive)) +
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  stat_pvalue_manual(stat.test, label = "{p.adj}")  + #add p value
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) + # Add 10% spaces between the p-value labels and the plot border
  geom_jitter()+
  ggtitle("opscc_hpv_score ")
plt
```


