---
title: '`r rstudioapi::getSourceEditorContext()$path %>% basename() %>% gsub(pattern = "\\.Rmd",replacement = "")`' 
author: "Avishai Wizel"
date: '`r Sys.time()`'
output: 
  html_notebook: 
    code_folding: hide
    toc: yes
    toc_collapse: yes
    toc_float: 
      collapsed: FALSE
    number_sections: true
    toc_depth: 2
---



# Functions

```{r warning=FALSE}
source_from_github(repositoy = "DEG_functions",version = "0.2.53")
source_from_github(repositoy = "HMSC_functions",version = "0.1.14",script_name = "functions.R")
source_from_github(repositoy = "cNMF_functions",version = "0.4.04",script_name = "cnmf_functions_V3.R")
source_from_github(repositoy = "sc_general_functions",version = "0.1.34",script_name = "functions.R")
library(hypeR)

```

# Data

```{r}
hmsc_cancer_cells = readRDS("./Data/acc1_cancer_cells_2500features_integrated_V5.RDS")
hmsc_acc_pri_cancer = readRDS("./Data/hmsc_acc_pri_cancer_processed_V6.RDS")
acc_pri = subset(hmsc_acc_pri_cancer,subset = patient.ident != "HMSC")
DefaultAssay(acc_pri) = "RNA"

all_cells = readRDS("./Data/acc_tpm_nCount_mito_no146_15k_with_ACC1_V2.RDS")
genesets_h  = getGmt("./Data/h.all.v7.0.symbols.pluscc.gmt")

# luminal_pathways = c("CHARAFE_BREAST_CANCER_LUMINAL_VS_BASAL_DN","CHARAFE_BREAST_CANCER_LUMINAL_VS_BASAL_UP","CHARAFE_BREAST_CANCER_LUMINAL_VS_MESENCHYMAL_DN","CHARAFE_BREAST_CANCER_LUMINAL_VS_MESENCHYMAL_UP","HUPER_BREAST_BASAL_VS_LUMINAL_DN","LIM_MAMMARY_LUMINAL_PROGENITOR_UP","SMID_BREAST_CANCER_LUMINAL_B_UP" )
# 
# # add luminal pathways
# luminal_gs = msigdbr(species = "Homo sapiens") %>% as.data.frame() %>% dplyr::filter(gs_name %in% luminal_pathways)%>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()
```




# HMSC vs ACC

## UMAP

```{r}
DimPlot(hmsc_acc_pri_cancer,group.by = "patient.ident",label = T)
```

```{r}
FeaturePlot(object = hmsc_acc_pri_cancer,features = c("MYB"),pt.size = 1)
FeaturePlot(object = hmsc_acc_pri_cancer,features = c("kaye_acc_score"),pt.size = 1)

```



```{r}
pdf("./Figures/kaye_acc_score_AllCancerCells.pdf")
FeaturePlot(object = all_acc_cancer_cells,features = c("kaye_acc_score"),pt.size = 1)
dev.off()
```

## enrichment analysis 

```{r fig.width=8, echo=TRUE,results='hide',fig.keep='all'}
hmsc_acc_pri_cancer = SetIdent(hmsc_acc_pri_cancer, value ="patient.ident")
acc_deg <-
  FindMarkers(
    hmsc_acc_pri_cancer,
    ident.1 = "HMSC",
    features = VariableFeatures(hmsc_acc_pri_cancer),
    densify = T,
    assay = "RNA",
    mean.fxn = function(x) {
      return(log(rowMeans(x) + 1, base = 2)) # change fun to calculate logFC in log space data (default to exponent data)
    }
  )

```


```{r fig.height=5, fig.width=13, results='asis'}
ranked_vec = acc_deg[,"avg_log2FC"]%>% setNames(rownames(acc_deg)) %>% na.omit() # make named vector
hyp_obj <-hypeR_fgsea(signature = ranked_vec,genesets = geneIds(genesets_h),up_only = F)

plt = hyp_dots(hyp_obj,merge = F)
plt1 = plt$up+ aes(size=nes,color = geneset)+ggtitle("Up in HMSC")+ 
  guides(
    size = guide_legend(title="NES",reverse=T),
    color = guide_legend(title="Geneset size",reverse=T)) +
  scale_colour_gradient(low = "black",high = "red")

plt2 = plt$dn+ aes(size=nes,color = geneset)+ggtitle("Down in HMSC")+scale_size(trans = 'reverse')+
    guides(
    size = guide_legend(title="NES",reverse=F),
    color = guide_legend(title="Geneset size",reverse=T)) +
  scale_colour_gradient(low = "black",high = "red")

plt1+plt2
```


## Cell cycle score  {.tabset}



```{r}
hallmark_name = "GO_MITOTIC_CELL_CYCLE"
acc_pri = FindVariableFeatures(object = acc_pri,nfeatures = 15000)
acc_pri = ScaleData(object = acc_pri,features = VariableFeatures(acc_pri))
geneIds= genesets_h[[hallmark_name]]@geneIds %>% intersect(VariableFeatures(acc_pri)) 
score <- apply(acc_pri@assays$RNA@scale.data[geneIds,],2,mean)
acc_pri=AddMetaData(acc_pri,score,hallmark_name)

hmsc_cancer_cells = FindVariableFeatures(object = hmsc_cancer_cells,nfeatures = 15000)
hmsc_cancer_cells = ScaleData(object = hmsc_cancer_cells,features = VariableFeatures(hmsc_cancer_cells))
geneIds= genesets_h[[hallmark_name]]@geneIds %>% intersect(VariableFeatures(hmsc_cancer_cells)) 
score <- apply(hmsc_cancer_cells@assays$integrated@scale.data[geneIds,],2,mean)
hmsc_cancer_cells=AddMetaData(hmsc_cancer_cells,score,hallmark_name)

```

```{r echo=TRUE, results='asis'}
acc_cc_scores = FetchData(object = acc_pri,vars = "GO_MITOTIC_CELL_CYCLE")
hmsc_cc_scores = FetchData(object = hmsc_cancer_cells,vars = "GO_MITOTIC_CELL_CYCLE")

distributions_plt = ggplot() +
  geom_density(aes(GO_MITOTIC_CELL_CYCLE, fill = "ACC"), alpha = .2, data = acc_cc_scores) +
  geom_density(aes(GO_MITOTIC_CELL_CYCLE, fill = "HMSC"), alpha = .2, data = hmsc_cc_scores) +
  scale_fill_manual(name = "Dataset", values = c(ACC = "red", HMSC = "green"))+ geom_vline(aes(xintercept=0.3),
            color="blue", linetype="dashed", size=1) +ggtitle("'GO_MITOTIC_CELL_CYCLE'  score distribution")

print_tab(plt = distributions_plt,title = "score distribution",subtitle_num = 3)

```


```{r echo=TRUE, results='asis'}
hmsc_cc_cells = (sum(hmsc_cancer_cells@meta.data[[hallmark_name]]> 0.3) /ncol(hmsc_cancer_cells)) %>% round(digits = 3)*100
acc_cc_cells = (sum(acc_pri@meta.data[[hallmark_name]]> 0.3)/ncol(acc_pri)) %>% round(digits = 3)*100
df = data.frame(Dataset = c("HMSC","ACC"), cycling_cells_percentage = c(hmsc_cc_cells,acc_cc_cells))
cycling_cells_plt = ggplot(data=df, aes(x=Dataset, y=cycling_cells_percentage)) +
  geom_text(aes(label=cycling_cells_percentage), vjust=0, color="black", size=3.5)+
  geom_bar(stat="identity")+ylab(" % cycling cells")+
  geom_bar(stat="identity", fill="steelblue")+
  theme_minimal() + ggtitle("Cycling cells count")

print_tab(plt = cycling_cells_plt,title = "# cycling cells",subtitle_num = 3)
```
```{r}
pdf(file = "./Figures/CC_distributions.pdf")
distributions_plt
dev.off()

pdf(file = "./Figures/cycling_cells.pdf")
cycling_cells_plt
dev.off()
```


```{r fig.height=8, fig.width=10, results='asis'}
print_tab(plt = 
            FeaturePlot(hmsc_acc_pri_cancer,features = c("MKI67","CDK1","MCM2","CDC20"))
          ,title = "CC genes",subtitle_num = 3)
```

## Cycling cells filtering {.tabset}
```{r warning=FALSE}
#add score to all acc cancer cells
# geneIds= genesets[[hallmark_name]]@geneIds %>% intersect(VariableFeatures(all_acc_cancer_cells,assay = "integrated")) 
# score <- apply(all_acc_cancer_cells@assays$integrated@scale.data[geneIds,],2,mean)

#add score to all acc cancer cells
cc_all = c(acc_pri$GO_MITOTIC_CELL_CYCLE, hmsc_cancer_cells$GO_MITOTIC_CELL_CYCLE) %>% as.data.frame()
hmsc_acc_pri_cancer=AddMetaData(hmsc_acc_pri_cancer,cc_all,hallmark_name)

#filter:
all_acc_cancer_cells_ccFiltered=hmsc_acc_pri_cancer[,hmsc_acc_pri_cancer@meta.data[[hallmark_name]]< 0.3]


min_threshold = min(hmsc_acc_pri_cancer$GO_MITOTIC_CELL_CYCLE)
max_threshold = max(hmsc_acc_pri_cancer$GO_MITOTIC_CELL_CYCLE)
```

```{r, results='asis'}
library(viridis)

print_tab(plt = FeaturePlot(object = hmsc_acc_pri_cancer,features = hallmark_name) + ggtitle("Before cc filtering") &
            scale_color_gradientn(colours = plasma(n = 10, direction = -1), limits = c(min_threshold, max_threshold))
          ,title = "Before CC filtering",subtitle_num = 3)


print_tab(plt = 
            FeaturePlot(object = all_acc_cancer_cells_ccFiltered,features = hallmark_name) + 
            ggtitle("After cc filtering") &
            scale_color_gradientn(colours = plasma(n = 10, direction = -1), limits = c(min_threshold, max_threshold))
          ,title = "After CC filtering" ,subtitle_num = 3)

```



## DEG

```{r fig.width=8, echo=TRUE, results='asis',fig.keep='all'}
all_acc_cancer_cells_ccFiltered = SetIdent(all_acc_cancer_cells_ccFiltered, value ="patient.ident")
all_acc_cancer_cells_ccFiltered = FindVariableFeatures(all_acc_cancer_cells_ccFiltered,nfeatures = 15000,assay = "integrated")
acc_deg <-
  FindMarkers(
    all_acc_cancer_cells_ccFiltered,
    ident.1 = "HMSC",
    features = VariableFeatures(all_acc_cancer_cells_ccFiltered,assay = "integrated"),
    densify = T,
    verbose = T,
    slot = "data",
    mean.fxn = function(x) {
      return(log(rowMeans(x) + 1,base = 2)) # change func to calculate logFC in log space data (default to exponent data)
    },
    assay = "RNA"
  )
```



```{r fig.height=5, fig.width=13, results='asis'}
ranked_vec = acc_deg[,"avg_log2FC"]%>% setNames(rownames(acc_deg)) %>% na.omit() # make named vector
hyp_obj <-hypeR_fgsea(signature = ranked_vec,genesets = geneIds(genesets_h),up_only = F)

plt = hyp_dots(hyp_obj,merge = F)
plt1 = plt$up+ aes(size=nes)+ggtitle("Up in HMSC")+ 
  guides(
    size = guide_legend(title="NES",reverse=T)) 

plt2 = plt$dn+ aes(size=nes)+ggtitle("Down in HMSC")+scale_size(trans = 'reverse')+
    guides(
    size = guide_legend(title="NES",reverse=F))

plt1+plt2

```




```{r}
pdf(file = "./Figures/ACC_vs_HMSC_GSEA.pdf",width = 13,height = 5)
plt1+plt2
dev.off()
```

```{r}
volcano_plot.2 = function(de_genes, top_genes_text=0, title = "" ,show_gene_names = NULL, ident1 = "",
                        ident2 = "" , fdr_cutoff = 0.05 , fc_cutoff = 1.3,
                        return_de_genes = F, show_legend = T) {
  library(ggrepel,quietly = T)
  library(dplyr,quietly = T)
  names_for_label = c(paste("Genes up in",ident1),paste("Genes up in",ident2))
  #color genes if there are over/under/same expressed
  de_genes$diffexpressed <- "Same" 
  de_genes$avg_log2FC[!is.finite(de_genes$avg_log2FC)] <- 1000 #remove infinite values
  de_genes$fdr = p.adjust(p = de_genes$p_val ,method = "fdr")
  # if log2Foldchange > fc_cutoff and pvalue < 0.05, set as "UP" 
  de_genes$diffexpressed[de_genes$avg_log2FC > log2(fc_cutoff) & de_genes$fdr < fdr_cutoff] <- names_for_label[1]
  
  # if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
  de_genes$diffexpressed[de_genes$avg_log2FC < (-log2(fc_cutoff)) & de_genes$fdr < fdr_cutoff] <- names_for_label[2]
  
  #set name labels for highest genes
  de_genes$delabel <- NA
  down_genes = de_genes$diffexpressed == names_for_label[1]
  up_genes = de_genes$diffexpressed == names_for_label[2]
  
  de_genes$delabel[up_genes][0:top_genes_text] <- rownames(de_genes)[up_genes][0:top_genes_text]
  de_genes$delabel[down_genes][0:top_genes_text] <- rownames(de_genes)[down_genes][0:top_genes_text]
  
  
  #Show genes that specify in show_gene_names
  if (!is.null(show_gene_names)){
    de_genes_index = match(show_gene_names,rownames(de_genes)) #indexes of de_genes that in show_gene_names
    de_genes_index <- de_genes_index[!is.na(de_genes_index)] #remove NA
    
    show_gene_names_index = show_gene_names %in% rownames(de_genes) #indexes of show_gene_names that in de_genes
    de_genes$delabel[de_genes_index] = show_gene_names [show_gene_names_index]
  }
  
  #colors for diff exp genes
  cols <- structure(c("green4", "red", "grey"), .Names = c(names_for_label[2],names_for_label[1], "Same"))
  
  title = paste(title,"Differential gene expression in", ident1,"vs", ident2)
  p = ggplot(data=de_genes, aes(x=avg_log2FC, y=-log10(p_val), col=diffexpressed, label=delabel)) + 
    geom_point() + 
    theme_minimal() +
    scale_color_manual(values=cols) +
    geom_text_repel(na.rm = T,box.padding = 1,max.overlaps = Inf,color = "blue") +
    xlab("Average log2 fold-change")+ 
    ylab("p-value")+ 
    scale_y_continuous(labels = function(x) {parse(text = paste0("10^-",x))})+
    guides(col=guide_legend(title=paste0("Significant DEG\n(FDR<",fdr_cutoff," ,|FC| >", fc_cutoff,")")))+
    {if(show_legend == F) theme(legend.position="none")}+
    {if(show_legend) ggtitle(title) }+
    theme(axis.text  = element_text( color="black", size=12),axis.title = element_text( color="black", size=12))
  
  if (return_de_genes == T){ return(de_genes)}
  else {return (p)}
  
}

```

```{r}
volcano_plt = volcano_plot.2(de_genes = acc_deg,fdr_cutoff = 0.05,fc_cutoff = 2, ident1 = "HMSC",ident2 = "ACC",top_genes_text = 0)
volcano_plt
```






```{r}
pdf("./Figures/volcano_plot_ACC_VS_HMSC.pdf")
volcano_plt
dev.off()



```





## CNV plot {.tabset}
```{r}
# create expression matrix of acc + normal cells + HMSC seurat integrated
acc_all_cells_noAcc1 = subset(all_cells, subset = patient.ident != "ACC1")
acc_expr = acc_all_cells_noAcc1@assays$RNA@data %>% as.data.frame()
hmsc_expr  = hmsc_cancer_cells@assays$integrated@data %>% as.data.frame()
acc_expr = acc_expr [ rownames(hmsc_expr),]
all_expr = cbind(acc_expr,hmsc_expr)

# create annotation
acc_annotation_integrated  = readRDS("./Data/inferCNV/infercnv_input/cell_types.RDS")
acc_annotation_integrated = acc_annotation_integrated[colnames(all_expr),,drop = F]
acc_annotation_integrated = acc_annotation_integrated %>% rownames_to_column("orig.ident")

#write expression and annotation
write.table(acc_annotation_integrated, "./Data/inferCNV/infercnv_input/acc_annotation_integrated.txt", append = FALSE,
            sep = "\t", dec = ".",row.names = FALSE, col.names = F)


write.table(all_expr, "./Data/inferCNV/infercnv_input/all.4icnv_integrated.txt", append = FALSE,
            sep = "\t", dec = ".",row.names = T, col.names = T)
```
```{r}
  
my_fun <- function(x){
  a <- 2
  b <- 5
  return(a*x+b)
}
my_fun(2)
# 9

body(my_fun)[[2]] <- substitute(a <- 3)

my_fun(2)
```


```{r}
trace(infercnv::run,edit = T) # to skip normalization, change to skip_past = 4
#(https://github.com/broadinstitute/infercnv/issues/346)
```

```{r}

infercnv_obj = CreateInfercnvObject(raw_counts_matrix="./Data/inferCNV/infercnv_input/all.4icnv_integrated.txt", 
                                    annotations_file="./Data/inferCNV/infercnv_input/acc_annotation_integrated.txt",
                                    delim="\t",gene_order_file="./Data/inferCNV/gencode_v19_gene_pos.txt"
                                    ,ref_group_names=c("CAF", "Endothelial", "WBC")) #groups of normal cells

infercnv_obj_default = infercnv::run(infercnv_obj, cutoff=1, out_dir='./Data/inferCNV/infercnv_intergrated_output',
                                     cluster_by_groups=T, plot_steps=FALSE,
                                     denoise=TRUE, HMM=FALSE, no_prelim_plot=TRUE,
                                     png_res=300)
untrace(infercnv::run)


```


```{r}
trace(infercnv:::get_group_color_palette ,edit = T) # change "Set3" to "Set1" for more distinguishable colors
plot_cnv(infercnv_obj_default, output_format = "png",  write_expr_matrix = FALSE,out_dir = "./Data/inferCNV/infercnv_intergrated_output",png_res	=800,obs_title = "Malignant cells",ref_title = "Normal cells",contig_cex = 2, title = "Copy number variation")
untrace(infercnv:::get_group_color_palette)
```

```{r fig.height=8, fig.width=8,results='asis'}
print_tab(plt = knitr::include_graphics("./Data/inferCNV/infercnv_intergrated_output/infercnv.png")
          ,title = "CNV plot",subtitle_num = 3)
```


```{r,results='asis'}
library(limma)
smoothed=apply(infercnv_obj_default@expr.data,2,tricubeMovingAverage, span=0.01)
cnsig=sqrt(apply((smoothed-1)^2,2,mean))

acc_all_cells <- AddMetaData(object = acc_all_cells, metadata = cnsig, col.name = "copynumber")
acc_all_cells = SetIdent(object = acc_all_cells,value = "cell.type")

print_tab(plt = FeaturePlot(acc_all_cells, "copynumber",pt.size = 1,label = T,repel = T)+
            scale_colour_gradientn(colours=c("white","lightblue","orange","red","darkred"))
          ,title = "CNV UMAP",subtitle_num = 3)

```




# HMSC analysis 
```{r}
DefaultAssay(hmsc_cancer_cells) = "integrated"
```

## UMAP 

```{r}
hmsc_cancer_cells = FindClusters(object = hmsc_cancer_cells,verbose = F,resolution = 0.5)
DimPlot(object = hmsc_cancer_cells,pt.size = 2)
```

## Scores 
```{r fig.height=8, fig.width=12}
FeaturePlot(object = hmsc_cancer_cells,features = c("MYB","JAG1"),pt.size = 2)+
DimPlot(object = hmsc_cancer_cells,group.by  = c("hpv33_positive"),pt.size = 2)

```





## NMF {.tabset}
```{python}
from cnmf import cNMF
import pickle
nfeatures = "2K"
f = open('./Data/cNMF/HMSC_cNMF_harmony_2Kvargenes/cnmf_obj.pckl', 'rb')
cnmf_obj = pickle.load(f)
f.close()
```



```{r}
knitr::include_graphics("./Data/cNMF/HMSC_cNMF_harmony_2Kvargenes/HMSC_cNMF_harmony_2Kvargenes.k_selection.png")
```

```{python}
selected_k = 3
density_threshold = 0.1
cnmf_obj.consensus(k=selected_k, density_threshold=density_threshold,show_clustering=True)
usage_norm, gep_scores, gep_tpm, topgenes = cnmf_obj.load_results(K=selected_k, density_threshold=density_threshold)
```

```{r}
gep_scores = py$gep_scores
gep_tpm = py$gep_tpm
all_metagenes= py$usage_norm
```

## Harmony results {.tabset}



```{r fig.height=10, fig.width=10, results='asis'}
# Make metagene names
for (i in 1:ncol(all_metagenes)) {
  colnames(all_metagenes)[i] = "metagene." %>% paste0(i)
}

#add each metagene to metadata
for (i in 1:ncol(all_metagenes)) {
  metage_metadata = all_metagenes %>% select(i)
  acc1_cancer_cells = AddMetaData(object = acc1_cancer_cells,metadata = metage_metadata)
}
print_tab(plt = 
            FeaturePlot(object = acc1_cancer_cells,features = colnames(all_metagenes),combine = T),
          title = "metagenes expression",subtitle_num = toc_tabs_level)

```


## Enrichment analysis by top 200 genes of each program
```{r fig.height=8, fig.width=8, results='hide'}

canonical_pathways = msigdbr(species = "Homo sapiens", category = "C2") %>% dplyr::filter(gs_subcat != "CGP") %>%  dplyr::distinct(gs_name, gene_symbol)

plt_list = list()
for (i in 1:ncol(gep_scores)) {
  top_genes = gep_scores  %>%  arrange(desc(gep_scores[i])) #sort by score a
  top = head(rownames(top_genes),200) #take top top_genes_num
  res = genes_vec_enrichment(genes = top,background = rownames(gep_scores),homer = T,title = 
                    i,silent = T,return_all = T,custom_pathways = canonical_pathways)
   
  plt_list[[i]] = res$plt
}
gridExtra::grid.arrange(grobs = plt_list)
```

```{r}
for (i in 1:ncol(gep_scores)) {
  ranked_vec = gep_scores %>% pull(i) %>%  setNames(rownames(gep_scores))
  hyp_obj <-hypeR_fgsea(signature = ranked_vec,genesets = genesets,up_only = T)

  plt = hyp_dots(hyp_obj,merge = F)+ aes(size=abs(nes))
  print(plt)
}
```


```{r fig.height=10}
library(ComplexHeatmap)
acc1_cancer_cells = SetIdent(object = acc1_cancer_cells,value = "seurat_clusters")
for (i in 1:ncol(gep_scores)) {
  top_genes = gep_scores  %>%  arrange(desc(gep_scores[i])) #sort by score a
  top = head(rownames(top_genes),50) #take top top_genes_num
  data = FetchData(object = acc1_cancer_cells,vars = top)%>% scale() %>% t() 
  metagene_data = FetchData(object = acc1_cancer_cells,vars = colnames(all_metagenes)[i])
  col_list = list(circlize::colorRamp2(c(0, 1), c("white", "red"))); names(col_list) = colnames(all_metagenes)[i]
  column_ha = HeatmapAnnotation(df = metagene_data,col = col_list)
  print(ComplexHeatmap::Heatmap(data,show_column_names = F,row_names_gp = grid::gpar(fontsize = 7),cluster_rows = F, top_annotation = 
                                  column_ha,name = "Z-score expression"))
    
  pdf(paste0("./Figures/NMF_top_genes_program",i,".pdf"))
  print(ComplexHeatmap::Heatmap(data,show_column_names = F,row_names_gp = grid::gpar(fontsize = 7),cluster_rows = F, top_annotation = 
                                  column_ha,name = "Z-score expression"))
  dev.off()
}

```






## Lum Myo score

```{r}
original_myo_genes = c( "TP63", "TP73", "CAV1", "CDH3", "KRT5", "KRT14", "ACTA2", "TAGLN", "MYLK", "DKK3")
original_lum_genes = c("KIT", "EHF", "ELF5", "KRT7", "CLDN3", "CLDN4", "CD24", "LGALS3", "LCN2", "SLPI" )
```

```{r fig.height=8, fig.width=12}
FeaturePlot(hmsc_cancer_cells,features = original_myo_genes)
FeaturePlot(hmsc_cancer_cells,features = original_lum_genes)

```


```{r}
acc_cancerCells_noACC1 = SetIdent(acc_cancerCells_noACC1,value = "patient.ident")
calculate_score(dataset = acc_cancerCells_noACC1,myo_genes = original_myo_genes,lum_genes = original_lum_genes)
```

```{r}
calculate_score.2 <- function(dataset,myo_genes,lum_genes,lum_threshold =1 , myo_threshold = -1) {
  myoscore=FetchData(object =dataset,vars =  myo_genes,slot = "data") %>% rowMeans()
  lescore=FetchData(object =dataset,vars =  lum_genes,slot = "data") %>% rowMeans()
  correlation = cor(lescore,myoscore) %>% round(digits = 2)
  message("correlation of lum score and myo score:" %>% paste(correlation))
  
  
  original_myo_genes = c("TP63", "TP73", "CAV1", "CDH3", "KRT5", "KRT14", "ACTA2", "TAGLN", "MYLK", "DKK3")
  original_lum_genes = c("KIT", "EHF", "ELF5", "KRT7", "CLDN3", "CLDN4", "CD24", "LGALS3", "LCN2", "SLPI")
  orig_myoscore=FetchData(object =dataset,vars =  original_myo_genes,slot = "data") %>% rowMeans()
  orig_lescore=FetchData(object =dataset,vars =  original_lum_genes,slot = "data") %>% rowMeans()
  correlation_to_original_lum = cor(orig_lescore,lescore) %>% round(digits = 2)
  correlation_to_original_myo = cor(orig_myoscore,myoscore) %>% round(digits = 2)

  message("correlation of lum score and original lum score:" %>% paste(correlation_to_original_lum))
  message("correlation of myo score and original myo score:" %>% paste(correlation_to_original_myo))

  dataset=AddMetaData(dataset,lescore-myoscore,"luminal_over_myo")
  print(
    FeaturePlot(object = dataset,features = "luminal_over_myo")
  )
  data = FetchData(object = dataset,vars = "luminal_over_myo")
  print(
    data %>% 
    ggplot(aes( x=luminal_over_myo)) + 
    geom_density() +ylab("Density")+theme(  axis.title=element_text(size=12,face="bold"))+ xlab("Luminal-Myoepithelial spectrum")
    )
  
lum_cells_num = subset(x = dataset,luminal_over_myo >(lum_threshold)) %>% ncol() /ncol(dataset)
myo_cells_num = subset(x = dataset,luminal_over_myo <(myo_threshold)) %>% ncol()/ncol(dataset)
df = data.frame(cell_type = c("myo_cells","lum_cells"),percentage = c(myo_cells_num,lum_cells_num))
ggplot(data=df, aes(x=cell_type, y=percentage)) +
  geom_bar(stat="identity") + ggtitle("ACC cell types")
}

```


## Original score of ACC1

```{r}
calculate_score.2(dataset = acc_pri,myo_genes = original_myo_genes,lum_genes = original_lum_genes,lum_threshold = 0,myo_threshold = 0)
```
```{r}
DefaultAssay(hmsc_cancer_cells) = "integrated"
```

```{r}
calculate_score(dataset = hmsc_cancer_cells,myo_genes = original_myo_genes,lum_genes = original_lum_genes,lum_threshold = 0,myo_threshold = 0)
```


```{r}
library(circlize)
data = FetchData(object = hmsc_cancer_cells,vars = c(original_lum_genes))
ComplexHeatmap::Heatmap(matrix = cor(data),name = "pearson", col = colorRamp2(seq(-1, 1, length = 3), c("blue", "#EEEEEE", "red")), heatmap_legend_param = list(at = c(-1, 0, 1)),
    cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.2f", cor(data)[i, j]), x, y, gp = gpar(fontsize = 9))
}, column_title = "HMSC luminal genes")

data = FetchData(object = acc_pri,vars = c(original_lum_genes))
ComplexHeatmap::Heatmap(matrix = cor(data),name = "pearson", col = colorRamp2(seq(-1, 1, length = 3), c("blue", "#EEEEEE", "red")), heatmap_legend_param = list(at = c(-1, 0, 1)),
    cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.2f", cor(data)[i, j]), x, y, gp = gpar(fontsize = 9))
}, column_title = "ACC luminal genes")
```
```{r}
library(circlize)
data = FetchData(object = hmsc_cancer_cells,vars = c(original_myo_genes))
ComplexHeatmap::Heatmap(matrix = cor(data),name = "pearson", col = colorRamp2(seq(-1, 1, length = 3), c("blue", "#EEEEEE", "red")), heatmap_legend_param = list(at = c(-1, 0, 1)),
    cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.2f", cor(data)[i, j]), x, y, gp = gpar(fontsize = 9))
}, column_title = "HMSC Myoepithelial genes")

data = FetchData(object = acc_pri,vars = c(original_myo_genes))
ComplexHeatmap::Heatmap(matrix = cor(data),name = "pearson", col = colorRamp2(seq(-1, 1, length = 3), c("blue", "#EEEEEE", "red")), heatmap_legend_param = list(at = c(-1, 0, 1)),
    cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.2f", cor(data)[i, j]), x, y, gp = gpar(fontsize = 9))
}, column_title = "ACC Myoepithelialgenes")
```

## HPV
### HPV UMAP

```{r}
HPV33_P3 = fread("./Data/HPV33_P3.txt",col.names = c("plate","reads")) %>% as.data.frame()
HPV33_P3.df = HPV33_P3 %>% mutate(
  plate = gsub(x =HPV33_P3$plate, replacement = "",pattern = "_.*$") 
  %>% gsub(pattern = "-P",replacement = ".P") 
  %>% gsub(pattern = "-",replacement = "_",)
  )
HPV33_P3.df = HPV33_P3.df %>% dplyr::filter(HPV33_P3.df$plate %in% colnames(hmsc_cancer_cells))
rownames(HPV33_P3.df)  <- HPV33_P3.df$plate
HPV33_P3.df$plate = NULL


HPV33_P2 = fread("./Data/HPV33_P2.txt",col.names = c("plate","reads")) %>% as.data.frame()
HPV33_P2.df = HPV33_P2 %>% mutate(
  plate = gsub(x =HPV33_P2$plate, replacement = "",pattern = "_.*$") 
  %>% gsub(pattern = "plate2-",replacement = "plate2_",)
  %>% gsub(pattern = "-",replacement = "\\.",)
  )
HPV33_P2.df = HPV33_P2.df %>% dplyr::filter(HPV33_P2.df$plate %in% colnames(hmsc_cancer_cells))
rownames(HPV33_P2.df)  <- HPV33_P2.df$plate
HPV33_P2.df$plate = NULL

HPV33 = rbind(HPV33_P3.df,HPV33_P2.df)
hmsc_cancer_cells = AddMetaData(object = hmsc_cancer_cells,metadata = HPV33,col.name = "HPV33.reads")
FeaturePlot(hmsc_cancer_cells,features = "HPV33.reads",max.cutoff = 10)
```


```{r}

data = FetchData(object = hmsc_cancer_cells,vars = "HPV33.reads")

data = data %>% mutate("0 reads" = if_else(condition = HPV33.reads == 0,true = 1,false = 0))
data = data %>% mutate("1 reads" = if_else(condition = HPV33.reads == 1,true = 1,false = 0))
data = data %>% mutate("2 reads" = if_else(condition = HPV33.reads == 2,true = 1,false = 0))
data = data %>% mutate("3-23 reads" = if_else(condition = HPV33.reads >=3 &HPV33.reads  <24,true = 1,false = 0))
data = data %>% mutate("24+ reads" = if_else(condition = HPV33.reads >=24,true = 1,false = 0))
data = colSums(data[,2:ncol(data)]) %>% as.data.frame()
names(data)[1] = "count"
data = rownames_to_column(data,var = "bin")
data
ggplot(data=data, aes(x=factor(bin,levels = c("0 reads","1 reads","2 reads","3-23 reads","24+ reads")), y=count)) +
  geom_bar(stat="identity", fill="steelblue") + xlab("HPV Reads")+ theme_minimal()+
  geom_text(aes(label=count), vjust=-0.5, color="black", size=3.5)
```


```{r}
hpv33_positive = HPV33 %>% dplyr::mutate(hpv33_positive = case_when(reads >= 10 ~ "positive",
                                                                    reads < 10 ~ "negative")
)



hpv33_positive$reads = NULL
hmsc_cancer_cells = AddMetaData(object = hmsc_cancer_cells,metadata = hpv33_positive)
```

```{r fig.height=6, fig.width=12}
DimPlot(object = hmsc_cancer_cells,group.by  = c("hpv33_positive"),pt.size = 2)+
FeaturePlot(object = hmsc_cancer_cells,features = "MYB",pt.size = 2)

```






### DEG LR latent vars plate
```{r}
DefaultAssay(hmsc_cancer_cells) = "integrated"
hmsc_cancer_cells = SetIdent(hmsc_cancer_cells,value = "hpv33_positive")
```



```{r}
library("biomaRt")

features = VariableFeatures(hmsc_cancer_cells)
features = hmsc_cancer_cells@assays$RNA@data %>% rowMeans() %>% sort(decreasing = T) %>% head(3000) %>% names()

acc_deg <-
  FindMarkers(
    hmsc_cancer_cells,
    ident.1 = "positive",
    ident.2 = "negative",
    features = features,
    densify = T,
    assay = "RNA",
    test.use = "LR",
    latent.vars = "plate",
    logfc.threshold = 0.1,
    min.pct = 0.1,
    only.pos = F,
    mean.fxn = function(x) {
      return(log(rowMeans(x) + 1, base = 2)) # change func to calculate logFC in log space data (default to exponent data)
    }
  )
acc_deg$fdr<-p.adjust(p = as.vector(acc_deg$p_val) ,method = "fdr" )

```


```{r}
ranked_vec = acc_deg[,"avg_log2FC"]%>% setNames(rownames(acc_deg)) %>% na.omit() # make named vector

hyp_obj <-hypeR_fgsea(signature = ranked_vec,genesets = genesets,up_only = F)
hyp_dots(hyp_obj,merge = F)×–

```

```{r}
acc_deg
acc_deg[c("MYB","JAG1"),]

```





volcano plot log2(mean logTPM HPV+) - log2(mean logTPM HPV-)
```{r}
plt = volcano_plot.2(de_genes = acc_deg,fc_cutoff = 1.3, fdr_cutoff = 0.1,show_gene_names = c("MYB","JAG1"),ident1 = "HPV33 positive",ident2 = "HPV33 negative",top_genes_text = 5) +ggtitle("Differential Gene Expression by HPV Status
")
plt
```

```{r}
pdf("./Figures/HMSC_HPVon_HPVoff_volcano.pdf")
pltw
dev.off()
```

volcano plot log2(mean logTPM HPV+) - log2(mean logTPM HPV-)

```{r}

acc_deg <-
  FindMarkers(
    hmsc_cancer_cells,
    ident.1 = "positive",
    ident.2 = "negative",
    features = features,
    densify = T,
    assay = "RNA",
    test.use = "wilcox",
    latent.vars = "plate",
    logfc.threshold = 0.1,
    min.pct = 0.1,
    mean.fxn = function(x) {
      return((rowMeans(x)+1)) # change func to calculate logFC in log space data (default to exponent data)
    }
  )
acc_deg$fdr<-p.adjust(p = as.vector(acc_deg$p_val) ,method = "fdr" )
```

```{r}
acc_deg[c("MYB","JAG1"),]
```


volcano plot (mean logTPM HPV+) - (mean logTPM HPV-)

```{r}
volcano_plot(de_genes = acc_deg,fc_cutoff = 1.3, fdr_cutoff = 0.1,show_gene_names = c("MYB","JAG1"),ident1 = "HPV33 positive",ident2 = "HPV33 negative",top_genes_text = 5)+xlab("avg diff")
```

### HPV vs genes
```{r fig.width=8}
top_genes = acc_deg %>% filter(.$avg_log2FC>0) %>% arrange(.$fdr) %>% head(5) %>% rownames()
top_genes = c(top_genes,c("MYB","JAG1"))

myb_vs_hpv = FetchData(object = hmsc_cancer_cells,vars = c("hpv33_positive",top_genes))
df = reshape2::melt(myb_vs_hpv,value.name = "Expression") %>% dplyr::rename(gene = variable)

library(rstatix)
stat.test <- df %>%
    group_by(gene) %>%
  wilcox_test(Expression ~ hpv33_positive) %>%
  mutate(y.position = 5)

stat.test

stat.test <- stat.test %>% 
  add_xy_position(x = "gene", dodge = 0.8)

ggboxplot(
  df,
  x = "gene",
  y = "Expression",
  color = "hpv33_positive",
  palette = "jco",
  add = "jitter"
)+ stat_pvalue_manual(stat.test,y.position = 13, label = "p = {p}",remove.bracket = T)
```

```{r}
metagenes_violin_compare.2 = function(dataset,prefix = "",pre_on = c("OSI","NT"),axis.text.x = 11,test = "t.test", programs = c("Hypoxia","TNFa","Cell_cycle"),return_list = F,combine_patients = F){
  require(facefuns)
  plt.lst = list()
  if(combine_patients){
    genes_by_tp = FetchData(object = dataset,vars =  c("treatment",programs)) %>% filter(treatment %in% pre_on)  %>% as.data.frame() #mean expression
    formula <- as.formula( paste("c(", paste(programs, collapse = ","), ")~ treatment ") )
    
    #plot and split by patient:   
    stat.test = compare_means(formula = formula ,data = genes_by_tp,method = test,p.adjust.method = "fdr")%>% # Add pairwise comparisons p-value
      dplyr::filter(group1 == pre_on[1] & group2 == pre_on[2])  #filter for pre vs on treatment only
    
    stat.test$p.format =stat.test$p.adj #modift 0 pvalue to be lowest possible float
    stat.test$p.format[!stat.test$p.format == 0 ] <- paste("=",stat.test$p.format[!stat.test$p.format == 0 ])
    stat.test$p.format[stat.test$p.format == 0 ] <- paste("<",.Machine$double.xmin %>% signif(digits = 3))
    
    
    genes_by_tp = reshape2::melt(genes_by_tp, id.vars = c("treatment"),value.name = "score")
    plt = ggplot(genes_by_tp, aes(x = variable, y = score,fill = treatment)) + geom_split_violin(scale = 'width')+ 
      geom_boxplot(width = 0.25, notch = FALSE, notchwidth = .4, outlier.shape = NA, coef=0)+
      ylim(min(genes_by_tp$score),max(genes_by_tp$score)*1.25)
    plt = plt +stat_pvalue_manual(stat.test, label = "{p.signif}", label.size = 7,  #add p value
                                  y.position = max(genes_by_tp$score)*1.08,inherit.aes = F,size = 3.3,x = ".y.") # set position at the top value
    return(plt)
  }
  
  
  
  
  
  if (return_list) {
    return(plt.lst)
  }
}
```

```{r fig.height=4, fig.width=7}
hmsc_cancer_cells$treatment = hmsc_cancer_cells$hpv33_positive %>% gsub(pattern = "negative",replacement = "HPV-negative")%>% gsub(pattern = "positive",replacement = "HPV-positive")
hmsc_cancer_cells@meta.data[["treatment"]] = factor(hmsc_cancer_cells$treatment, levels = c("HPV-positive", "HPV-negative"))


plt = metagenes_violin_compare.2(dataset = hmsc_cancer_cells, prefix = "patient",pre_on = c("HPV-positive","HPV-negative"),test = "wilcox.test",programs = top_genes, return_list = F,combine_patients = T)+ labs(fill = "")+ylab("Expression")+xlab("Gene")+theme(axis.text=element_text(size=12),
        axis.title=element_text(size=14))+ labs(fill = "")+ylab("Expression")+xlab("Gene")+theme(axis.title=element_text(size=14))+ scale_fill_manual(values=c("#F8766D", "cyan3")) #from bovona.rmd
plt
```

```{r}
pdf("./Figures/HMSC_HPVon_HPVoff_topGenes_violin.pdf")
plt
dev.off()
```

